### PR TITLE
Sketcher: Fix symmetry of slot redundancy false positive

### PR DESCRIPTION
**Pull request by PaddleStroke**
Perturb symmetric geometries when using the 'create symmetric constraints' option to avoid numerical singularities in the solver (Jacobian Rank). If geometry is "perfect", the solver cannot distinguish between the derivative of a Symmetry constraint and an Equal constraint, flagging one as redundant.

Fix https://github.com/FreeCAD/FreeCAD/issues/13551

See [this comment](https://github.com/FreeCAD/FreeCAD/issues/13551#issuecomment-3705421604) explaining why the bug was happening.

To fix the issue, I propose that we introduce a small perturbation to the arcs angle when the option 'create symmetric constraints' is enabled.
With this the solver correctly solves and does not report the false positive redundancy.

Please squash

### CODE REVIEW

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 1]
```diff
@@ -4390,6 +4390,22 @@ int SketchObject::addSymmetric(const std::vector<int>& geoIdList, int refGeoId,
 
     std::vector<Part::Geometry*> symgeos = getSymmetric(geoIdList, geoIdMap, isStartEndInverted, refGeoId, refPosId);
 
+    // Perturb geometry to avoid numerical singularities in the solver (Jacobian Rank).
+    // If geometry is "perfect", the solver cannot distinguish between the derivative
+    // of a Symmetry constraint and an Equal constraint, flagging one as redundant.
+    // see https://github.com/FreeCAD/FreeCAD/issues/13551
+    if (addSymmetryConstraints) {
+        for (auto* geo : symgeos) {
+            if (auto* arc = static_cast<Part::GeomArcOfCircle*>(geo)) {
```
Comment (kadet1090) [inline src/Mod/Sketcher/App/SketchObject.cpp]: Shouldn't it be dynamic_cast? `static_cast` will always return you something, unless `geo` is nullptr before cast - but in that case we should check that before checking cast as having static_cast in conditoin suggest that we expect the cast to fail, not geo to be nullptr.

Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp]: yes of course!

Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp]: Thanks for the catch.

**Comment by yorikvanhavre** [comment]
@AjinkyaDahale could you have a look at this one?

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 2]
```diff
@@ -4390,6 +4390,22 @@ int SketchObject::addSymmetric(const std::vector<int>& geoIdList, int refGeoId,
 
     std::vector<Part::Geometry*> symgeos = getSymmetric(geoIdList, geoIdMap, isStartEndInverted, refGeoId, refPosId);
 
+    // Perturb geometry to avoid numerical singularities in the solver (Jacobian Rank).
```
Comment (AjinkyaDahale) [inline src/Mod/Sketcher/App/SketchObject.cpp:4393]: Is this coming from any literature or a fresh invention?

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 3]
```diff
@@ -4390,6 +4390,22 @@ int SketchObject::addSymmetric(const std::vector<int>& geoIdList, int refGeoId,
 
     std::vector<Part::Geometry*> symgeos = getSymmetric(geoIdList, geoIdMap, isStartEndInverted, refGeoId, refPosId);
 
+    // Perturb geometry to avoid numerical singularities in the solver (Jacobian Rank).
+    // If geometry is "perfect", the solver cannot distinguish between the derivative
+    // of a Symmetry constraint and an Equal constraint, flagging one as redundant.
+    // see https://github.com/FreeCAD/FreeCAD/issues/13551
+    if (addSymmetryConstraints) {
+        for (auto* geo : symgeos) {
+            if (auto* arc = dynamic_cast<Part::GeomArcOfCircle*>(geo)) {
```
Comment (AjinkyaDahale) [inline src/Mod/Sketcher/App/SketchObject.cpp:4400]: Will this not affect arcs of ellipses and/or other conics?

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 4]
```diff
@@ -4390,6 +4390,22 @@ int SketchObject::addSymmetric(const std::vector<int>& geoIdList, int refGeoId,
...
+    // If geometry is "perfect", the solver cannot distinguish between the derivative
+    // of a Symmetry constraint and an Equal constraint, flagging one as redundant.
+    // see https://github.com/FreeCAD/FreeCAD/issues/13551
+    if (addSymmetryConstraints) {
+        for (auto* geo : symgeos) {
+            if (auto* arc = dynamic_cast<Part::GeomArcOfCircle*>(geo)) {
+                double start, end;
+                double perturbation = 0.001;
+                arc->getRange(start, end, true);
+                arc->setRange(start + perturbation, end, true);
```
Comment (AjinkyaDahale) [inline src/Mod/Sketcher/App/SketchObject.cpp]: Is there a case where end would have to be perturbed rather than start? Like starting with the reflected geometry, say?

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 3 above)
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp:4400]: Indeed it is only tackling arc of circles. Since the other arcs are kind of rare I have not added a fix for them

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 4 above)
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp]: Hmm. There is the edge case were end-start < perturbation. In which case we would setRange with a start bigger than the end.
It does not sound very likely but we can add a small test just to make sure.
So perhaps we should rather do arc->setRange(start, end + perturbation, true);

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 3 above)
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/App/SketchObject.cpp:4400]: It's still easy enough to replace this with an `isDerivedFrom`, if you can confirm this issue happens for them.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 2 above)
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp:4393]: It's gemini that explained that to me https://github.com/FreeCAD/FreeCAD/issues/13551#issuecomment-3705421604

And perturbing the arcs do solve the bug.

**Comment by PaddleStroke** [comment]
I just tested after rebasing, and it seems the functionality "add symmetric constraint' is broken, symmetric constraints are not being added anymore. At least in my branch.
I'll investigate tomorrow

**Comment by AjinkyaDahale** [comment]
> I just tested after rebasing, and it seems the functionality "add symmetric constraint' is broken, symmetric constraints are not being added anymore. At least in my branch. I'll investigate tomorrow

Oof, I was hoping this is not the case. This is likely because of the refactor. I'll take a look.

**Comment by AjinkyaDahale** [comment]
See https://github.com/FreeCAD/FreeCAD/pull/26886. The bug turned out to be somewhere else, but still an oversight on my end.

**Comment by AjinkyaDahale** [comment]
Is ellipse movement currently broken? When I try to move an ellipse by its edge, the centre instead jumps to the cursor first. If not reported already, I'll create the issue.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 3 above)
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp:4400]: Indeed this is right. We can test a slot with arc of ellipse, aop aoh and if the same happens then I can add the same fix for them as well.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 4 above)
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp]: Let's try `arc->setRange(start + perturbation, end + perturbation, true)` then to make sure it's ok.
I'll test when your fix gets in.

**Comment by PaddleStroke** [comment]
@AjinkyaDahale the same happens in 0.21 when dragging an ellipse. So not a regression. I don't remember an existing issue about it so you can create one.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 4 above)
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp]: I just tested and `arc->setRange(start + perturbation, end + perturbation, true)` doesn't work. I guess we need to perturbe the arc angle.
`arc->setRange(start + perturbation, end, true)` works great even if the arc angle is less than perturbation (which is like 0.05deg so very very small anyway).

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 3 above)
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp:4400]: The bug does not happen with other arcs.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 4 above)
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/App/SketchObject.cpp]: This requires further exploration then, because this issue is possible to come up again in a different form otherwise. `arc->setRange(start + perturbation, end + perturbation, true)` should have caused the same kind of error for the solver to fix, intuitively.

That said, it shouldn't block this PR being merged, since we are providing the correct constraints even if the starting point is slightly off.

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 5]
```diff
@@ -4390,6 +4390,22 @@ int SketchObject::addSymmetric(const std::vector<int>& geoIdList, int refGeoId,
...
 
+    // Perturb geometry to avoid numerical singularities in the solver (Jacobian Rank).
+    // If geometry is "perfect", the solver cannot distinguish between the derivative
+    // of a Symmetry constraint and an Equal constraint, flagging one as redundant.
+    // see https://github.com/FreeCAD/FreeCAD/issues/13551
+    if (addSymmetryConstraints) {
+        for (auto* geo : symgeos) {
+            if (auto* arc = dynamic_cast<Part::GeomArcOfCircle*>(geo)) {
+                double start, end;
+                double perturbation = 0.001;
```
Comment (AjinkyaDahale) [inline src/Mod/Sketcher/App/SketchObject.cpp]: It may be better to use `perturbation = Precision::Angular()` or `perturbation = 2 * Precision::Angular()` or some other multiple.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 3 above)
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/App/SketchObject.cpp:4400]: Please mention that in the comment as well then, for future reference.

**Comment by AjinkyaDahale** [comment]
Once the `perturbation` hard-coding is removed and the comment about other conics is added, I have no further issues.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 5 above)
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp]: good idea. tested and it works

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 3 above)
Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp:4400]: added

**Comment by chennes** [comment]
Thanks for the review, @AjinkyaDahale (and of course for the PR, @PaddleStroke!).

**Comment by freecad-ci-runner** [comment]
Successfully created backport PR for `releases/FreeCAD-1-1`:
- #26917
